docs/development/pr-review.asciidoc

```diff
@@ -0,0 +1,129 @@
+
+All of our code should have unit tests that verify its behaviors, including not only the "happy path", but also edge cases, error handling, etc.  When you change an existing API of a module, then there should always be at least one failing unit test, which in turn means we need to verify that all code consuming that API properly handles the change if necessary. For modules at a high enough level, this will mean we have breaking change in the product, which we'll need to handle accordingly.
+
+In addition to extensive unit test coverage, PRs should include relevant functional and integration tests. In some cases, we may simply be testing a programmatic interface (e.g. a service) that is integrating with the file system, the network, Elasticsearch, etc. In other cases, we'll be testing REST APIs over HTTP or comparing screenshots/snapshots with prior known acceptable state. In the worst case, we are doing browser-based functional testing on a running instance of Kibana using selenium.
+
+Enhancements are pretty much always going to have extensive unit tests as a base as well as functional and integration testing. Bug fixes should always include regression tests to ensure that same bug does not manifest again in the future.
+--
+
+
+[float]
+=== Product level review
+
+Reviewers are not simply evaluating the code itself, they are also evaluating the quality of the user-facing change in the product. This generally means they need to check out the branch locally and "play around" with it. In addition to the "do we want this change in the product" details from above, the reviewer should be looking for bugs and evaluating how approachable and useful the feature is as implemented. Special attention should be given to error scenarios and edge cases to ensure they are all handled well within the product.
+
+
+[float]
+=== Consistency, style, readability
+
+Having a relatively consistent codebase is an important part of us building a sustainable project. If 40+ different developers all wrote code in whatever style they personally desired, the resulting inconsistencies would make working across the codebase more challenging. We know this because it's happened before in the Kibana project with far fewer developers than we have contributing to the project today, and the process of retroactively getting the codebase into a consistent state was extremely time intensive and disruptive to all contributors.
+
+On the other hand, improvements are being added to JavaScript, TypeScript, and our dependencies all the time, so we should be flexible about evolving our style and coding standards over time. Absolute consistency across the whole codebase is not our goal here - instead, we seek to move together with consensus as we evolve our codebase over time.
+
+As always, automation is the best way we can help ensure consistency, so we enforce a pretty comprehensive set of linting rules through CI. We're also rolling out prettier to make this even more automatic.
```

Comment:
Nitpick: I think the three previous paragraphs can be shortened and merged into one. It's a bit "wordy" right now imo.

Reply:
Merged them together